### PR TITLE
[WIP] Rqlite

### PR DESCRIPTION
**Pull request by mkorszun**

### CODE REVIEW

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+---
+layout: post
+title: rqlite - building SQLite clusters
+date: 2015-03-14 00:00:00
+description:
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: Short description, shorter than 140 chars?

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+hn-discussion:
+tags:
+- raft
+- sqlite
+- iot
+- paxos
+- distributed consensus
+---
+
+Today our compass lead us into direction of [rqlite](https://github.com/otoolep/rqlite). It a is small, comprehensive, [Raft](https://raftconsensus.github.io/) based clustering middleware for [SQLite](https://sqlite.org/) databases. Someone could ask who the hell would need SQLite cluster but first things first.
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: I think it's better if you break down this in multiple lines, easier to read on text editor. Same for other paragraphs in the post.

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: into the direction...? Not sure.

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+Rqlite is using Raft [consensus protocol](http://en.wikipedia.org/wiki/Consensus_%28computer_science%29) to perform replication among SQLite database nodes. In short Raft is an alternative to [Paxos](http://en.wikipedia.org/wiki/Paxos_%28computer_science%29) which is hardly understood among population including ourselves. In general Raft ensures that if any state machine applies some operation as the `nth command`, no other state machine will ever apply a different `nth command`. This leads to consistency among distributed state machines and this is something Rqlite guaranties among SQLite nodes - consistency and partition tolerance - CP among [CAP](http://de.wikipedia.org/wiki/CAP-Theorem).
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: If project name is rqlite I wouldn't put the R in capital letter.

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: alternative to Paxos, (comma) which is...

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+Now we know what rqlite is doing and more or less how. But lets answer most important question - who and why would need it? To answer above questions lets first go back to official SQLite manifesto:
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: let's

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+***SQLite is a good fit for use in cellphones, set-top boxes, televisions, game consoles, cameras, watches, kitchen appliances, thermostats, automobiles, machine tools, airplanes, remote sensors, drones, medical devices, and robots: the "internet of things"***
+
+Exactly, kitchen appliances, thermostats, remote sensor, drones, robots - Internet Of Thins - is it not what is hot right now and according to many predictions will be next big thing? We bet it will.
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: of Things

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+Now we know what rqlite is doing and more or less how. But lets answer most important question - who and why would need it? To answer above questions lets first go back to official SQLite manifesto:
+
+***SQLite is a good fit for use in cellphones, set-top boxes, televisions, game consoles, cameras, watches, kitchen appliances, thermostats, automobiles, machine tools, airplanes, remote sensors, drones, medical devices, and robots: the "internet of things"***
+
+Exactly, kitchen appliances, thermostats, remote sensor, drones, robots - Internet Of Thins - is it not what is hot right now and according to many predictions will be next big thing? We bet it will. 
+
+Currently while implementing IoT solutions you usually use some cloud based data feed and sync mechanisms, i.e [firebase](https://www.firebase.com/), [relayr](https://relayr.io/) but sooner or later your sensors will need to share some data/state only among each other without need to synchronise it over the public network which is faster, cheaper and more reliable.
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: , (comma) but sonner...

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: , (comma) without the need of synchronizing it over...

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: , (comma) which is faster...

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+Exactly, kitchen appliances, thermostats, remote sensor, drones, robots - Internet Of Thins - is it not what is hot right now and according to many predictions will be next big thing? We bet it will. 
+
+Currently while implementing IoT solutions you usually use some cloud based data feed and sync mechanisms, i.e [firebase](https://www.firebase.com/), [relayr](https://relayr.io/) but sooner or later your sensors will need to share some data/state only among each other without need to synchronise it over the public network which is faster, cheaper and more reliable. 
+
+SQLite due to its simplicity is perfect to be run on almost any type of connected device. Together with rqlite it can give IoT makers solution to sync any type of data locally. Due to this opportunity we strongly believe that rqlite is a very promising, still fresh but active project. Among that it is implemented in [Go](https://golang.org/) another big thingy in programming languages world. It is really hard to find good reason to not contribute to this small but already interesting project. And we plan to do it. Lets see how far we can go. Will keep you updated.
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: Due to its simplicity, SQLite is perfect...

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: Due to this opportunity, (comma) we strongly...

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: implemented in Go, (comma) another new big thing in the world of programming languages.

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: hard to find good reason**s** not to contribute...

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: Let's see...

Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: We'll keep you...

**Comment by fern4lvarez** [comment]
OK, maybe too many comments, but I like it a lot, quite motivational! :tada:

**Comment by mkorszun** [comment]
@fern4lvarez but what is you general feeling - is it they way to go? Maybe too little implementation details of the project or i have missed list of things we would like to improve?

**Comment by fern4lvarez** [comment]
Better wording (which I cannot improve much) apart, I like it as a first introduction to the project. We can write more technical posts in the future, as we're getting more into the project, or tackling specific problems.

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+rqlite is using Raft [consensus protocol] to perform replication among SQLite database nodes.
+In short Raft is an alternative to [Paxos], which is hardly understood among population including ourselves.
+In general Raft ensures that if any state machine applies some operation as the `nth command`, no other state machine will ever apply a different `nth command`.
+This leads to consistency among distributed state machines and this is something Rqlite guaranties among SQLite nodes - consistency and partition tolerance - CP among [CAP].
+
+Now we know what rqlite is doing and more or less how.
+But let's answer most important question - who and why would need it?
+To answer above questions lets first go back to official SQLite manifesto:
+
+***SQLite is a good fit for use in cellphones, set-top boxes, televisions, game consoles, cameras, watches, kitchen appliances, thermostats, automobiles, machine tools, airplanes, remote sensors, drones, medical devices, and robots: the "internet of things"***
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: Line too long

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+Exactly, kitchen appliances, thermostats, remote sensor, drones, robots - Internet Of Things - is it not what is hot right now and according to many predictions will be next big thing?
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: Line too long

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+We bet it will.
+
+Currently while implementing IoT solutions you usually use some cloud based data feed and sync mechanisms, i.e [firebase], [relayr], but sooner or later your sensors will need to share some data/state only among each other, without need to synchronise it over the public network, which is faster, cheaper and more reliable.
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: Line too long

**Comment by mkorszun** [comment]
@fern4lvarez should we ask Liz for some quick check?

**Comment by fern4lvarez** [comment]
I'll ping her and see if she finds some time for this.

**Comment by otoolep** [comment]
You do realise that for rqlite to work, it requires that a quorum of nodes must be contactable at all times, or the cluster cannot commit changes? So by IoT applications, if you mean systems comprised of many nodes (100s) or where many nodes are not contactable all the time, rqlite may not be suitable?

**Comment by mkorszun** [comment]
@otoolep Thank you very much much for sharing this opinion. Yes, rqlite is a CP system, and yes, probably it will fail as a general purpose mechanism for sharing / syncing data between hundreds or thousands of sensors. Still see some use cases for IoT, especially mission critical ones where consistency might be the main requirement. For example some group of sensors (few / dozen) which work together based on some consistent state / data shared among each other. Data between them needs to be replicated and consistent, so others can immediately take over the responsibility from the sensor which died (warm standby) - as an example. This is what I have in my mind at the moment - does it make sense?

In general we are in the early stage of playing with rqlite so right now our ideas are quite theoretical. So once again we are glad, that you as an author give us some feedback. Wondering what would be your use cases for the rqlite except the one which was the reason to implement it?

**Comment by otoolep** [comment]
OK, that makes sense. There are some questions about well the Raft consensus protocol will scale, past say 9 or 11 nodes. It may become too slow.

My use case was on a 3-to-5 node cluster of machines, to maintain a small amount of state with relational constraints.

**Comment by fern4lvarez** [comment]
Fix commit history, edit the date (so it appears as published today) and I'll merge.

File: _posts/2015-03-14-rqlite---building-SQLite-clusters.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
+---
+layout: post
+title: rqlite - building SQLite clusters
```
Comment (fern4lvarez) [inline _posts/2015-03-14-rqlite---building-SQLite-clusters.md]: What about "Building SQLite clusters with rqlite"? It looks better for a blog post title, doesn't it?

**Comment by mkorszun** [comment]
@fern4lvarez agree - will update it.

**Comment by mkorszun** [comment]
@fern4lvarez done

**Pull request merged**